File: python/src/zells/TurtleZell.py
import math


class TurtleZell(object):
    def __init__(self, name, emit):
        self.name = name
        self.emit = emit

        self.position = [0, 0]
        self.angle = 90
        self.segments = []
        self.canvases = {}

        print "Created Turtle " + name

    def receive(self, signal):
        if signal['to'] == '?':
            self.emit({
                'to': signal['from'],
                'content': [{
                    'to': self.name,
                    'from': signal['from'],
                    'content': '?'
                }]
            })

        if signal['to'] != self.name:
            return

        if '?' == signal['content']:
            self.emit({
                'to': signal['from'],
                'content': [
                    {
                        "to": self.name,
                        "content": {
                            "drawOn": "a display",
                            "size": {"width": 400, "height": 200}
                        }
                    }, {
                        'to': self.name,
                        'content': {'go': 'forward'}
                    }, {
                        'to': self.name,
                        'content': {'turn': 'left'}
                    }, {
                        'to': self.name,
                        'content': {'turn': 'right'}
                    }, {
                        'to': self.name,
                        'content': 'reset'
                    }]
            })

        if 'drawOn' in signal['content']:
            self.canvases[signal['content']['drawOn']] = signal['content']['size']
            self.draw()

        if 'go' in signal['content']:
            first = [self.position[0], self.position[1]]
            self.position[0] += math.cos(math.radians(self.angle))
            self.position[1] += math.sin(math.radians(self.angle))
            self.segments.append([first, [self.position[0], self.position[1]]])
            self.draw()

        if 'turn' in signal['content']:
            if signal['content']['turn'] == 'right':
                self.angle -= 15
            else:
                self.angle += 15
            self.draw()

        if 'reset' == signal['content']:
            self.position = [0, 0]
            self.angle = 90
            self.segments = []
            self.draw()

    def draw(self):
        for canvas in self.canvases:

            size = self.canvases[canvas]
            stepX = size['width'] / 100
            stepY = size['height'] / 100

            point1x = round(self.position[0] * stepX + math.cos(math.radians(self.angle + 90)) * 10, 2)
            point1y = round(self.position[1] * stepY + math.sin(math.radians(self.angle + 90)) * 10, 2)
            point2x = round(self.position[0] * stepX + math.cos(math.radians(self.angle - 90)) * 10, 2)
            point2y = round(self.position[1] * stepY + math.sin(math.radians(self.angle - 90)) * 10, 2)
            point3x = round(self.position[0] * stepX + math.cos(math.radians(self.angle + 0)) * 20, 2)
            point3y = round(self.position[1] * stepY + math.sin(math.radians(self.angle + 0)) * 20, 2)

            strokes = []
            strokes.append('clear')

            for segment in self.segments:
                strokes.append({'line': {'from': {'x': round(segment[0][0] * stepX, 2), 'y': round(segment[0][1] * stepY, 2)},
                                         'to': {'x': round(segment[1][0] * stepX, 2), 'y': round(segment[1][1] * stepY, 2)},
                                         'color': {'r': 1, 'g': 1, 'b': 1},
                                         'width': 1}})

            strokes.append({'line': {'from': {'x': point1x, 'y': point1y},
                                     'to': {'x': point2x, 'y': point2y},
                                     'color': {'r': 0, 'g': 1, 'b': 0},
                                     'width': 1}})
            strokes.append({'line': {'from': {'x': point2x, 'y': point2y},
                                     'to': {'x': point3x, 'y': point3y},
                                     'color': {'r': 0, 'g': 1, 'b': 0},
                                     'width': 1}})
            strokes.append({'line': {'from': {'x': point3x, 'y': point3y},
                                     'to': {'x': point1x, 'y': point1y},
                                     'color': {'r': 0, 'g': 1, 'b': 0},
                                     'width': 1}})

            self.emit({
                'to': canvas,
                'content': {'draw': {'strokes': strokes}}})

File: java/dish/src/main/java/org/zells/dish/zells/TurtleZell.java
package org.zells.dish.zells;

import org.zells.dish.codec.Codec;
import org.zells.dish.codec.impl.FlatByteTreeCodec;
import org.zells.dish.codec.impl.SignalByteSource;
import org.zells.dish.codec.impl.SignalTreeCodec;
import org.zells.dish.core.Dish;
import org.zells.dish.core.Signal;
import org.zells.dish.core.Zell;
import org.zells.dish.core.impl.StandardSignal;

import java.util.*;

public class TurtleZell implements Zell {

  private Codec codec = new FlatByteTreeCodec();
  private SignalTreeCodec translator = new SignalTreeCodec();
  private Dish dish;
  private Map<String, Map> canvases = new HashMap<>();

  private String name;
  private Point position = new Point(0, 0);
  private double angle = 0;

  public TurtleZell(Dish dish, String name) {
    this.dish = dish;
    this.name = name;
  }

  @Override
  public void receive(Signal signal) {
    Object decoded = codec.decode(new SignalByteSource(signal));
    Map<Object, Object> message = translator.asMap(decoded);

    if (!message.containsKey("to") || !message.containsKey("content")) {
      System.out.println("Not a message");
      return;
    }

    if (translator.asString(message.get("to")).equals("?")) {
      transmit(new Dictionary()
          .with("to", message.get("from"))
          .with("from", name)
          .with("content", Collections.singletonList(
              new Dictionary()
                  .with("to", name)
                  .with("from", message.get("from"))
                  .with("content", "?"))));
      return;
    }

    if (!translator.asString(message.get("to")).equals(name)) {
      System.out.println("Not for me");
      return;
    }


    if (translator.asString(message.get("content")).equals("?")) {
      transmit(new Dictionary()
          .with("to", message.get("from"))
          .with("from", name)
          .with("content", Arrays.asList(
              new Dictionary()
                  .with("to", name)
                  .with("content", new Dictionary()
                      .with("draw on", "a display")
                      .with("size", new Dictionary()
                          .with("width", 400)
                          .with("height", 200))),
              new Dictionary()
                  .with("to", name)
                  .with("content", new Dictionary()
                      .with("go", "forward")),
              new Dictionary()
                  .with("to", name)
                  .with("content", new Dictionary()
                      .with("turn", "left")),
              new Dictionary()
                  .with("to", name)
                  .with("content", new Dictionary()
                      .with("turn", "right"))
          )));
    }

    Map<Object, Object> content = translator.asMap(message.get("content"));

    if (content.containsKey("draw on")) {
      canvases.put(translator.asString(content.get("draw on")),
          translator.asMap(content.get("size")));
      draw();
    } else if (content.containsKey("go")) {
      position = position.translated(new Point(0, 10).rotated(angle));
      System.out.println("Position " + position.x + "," + position.y);
      draw();
    } else if (content.containsKey("turn")) {
      if (translator.asString(content.get("turn")).equals("left")) {
        angle += 15;
      } else {
        angle -= 15;
      }
      draw();
    } else if (translator.asString(message.get("content")).equals("reset")) {
      position = new Point(0, 0);
      angle = 0;
      draw();
    }
  }

  private void draw() {
    for (String canvas : canvases.keySet()) {
      ArrayList<Point> points = new ArrayList<>();
      points.add(position.translated(new Point(-10, 0).rotated(angle)));
      points.add(position.translated(new Point(10, 0).rotated(angle)));
      points.add(position.translated(new Point(0, 20).rotated(angle)));

      ArrayList<Line> lines = new ArrayList<>();
      lines.add(new Line(points.get(0), points.get(1)));
      lines.add(new Line(points.get(1), points.get(2)));
      lines.add(new Line(points.get(2), points.get(0)));

      List<Object> strokes = new ArrayList<>();
      strokes.add("clear");

      for (Line l : lines) {
        strokes.add(new Dictionary()
            .with("line", new Dictionary()
                .with("width", 2)
                .with("color", new Dictionary().with("r", 0).with("g", 1).with("b", 0))
                .with("from", new Dictionary().with("x", l.from.x).with("y", l.from.y))
                .with("to", new Dictionary().with("x", l.to.x).with("y", l.to.y))));
      }

      transmit(new Dictionary()
          .with("to", canvas)
          .with("from", name)
          .with("content", new Dictionary()
              .with("draw", new Dictionary()
                  .with("strokes", strokes
                  ))
          ));
    }
  }

  private void transmit(Object signal) {
    dish.transmit(StandardSignal.from(codec.encode(translator.translate(signal))));
  }

  private class Dictionary extends HashMap<Object, Object> {
    Dictionary with(Object key, Object value) {
      put(key, value);
      return this;
    }
  }

  private class Point {
    double x;
    double y;

    Point(double x, double y) {
      this.x = Math.round(x * 100) / 100;
      this.y = Math.round(y * 100) / 100;
    }

    Point translated(Point d) {
      return new Point(x + d.x, y + d.y);
    }

    Point rotated(double angle) {
      double radians = Math.toRadians(angle);
      return new Point(
          x * Math.cos(radians) - y * Math.sin(radians),
          y * Math.cos(radians) + x * Math.sin(radians));
    }
  }

  private class Line {
    Point from;
    Point to;

    Line(Point from, Point to) {
      this.from = from;
      this.to = to;
    }
  }
}
